File: sdks/cpp/tests/test_protocol.cpp
#include <gtest/gtest.h>

#include "croupier/sdk/protocol.h"

#include <vector>

namespace croupier {
namespace sdk {
namespace protocol {
namespace test {

class ProtocolTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// Test PutMsgID and GetMsgID
TEST_F(ProtocolTest, MsgIDEncoding) {
    uint8_t buf[3];

    // Test various MsgID values
    std::vector<uint32_t> test_ids = {
        MSG_INVOKE_REQUEST,
        MSG_INVOKE_RESPONSE,
        MSG_REGISTER_REQUEST,
        MSG_HEARTBEAT_RESPONSE,
        0x010101,
        0xFFFFFF,
        0x000001,
        0x123456
    };

    for (uint32_t original_id : test_ids) {
        PutMsgID(buf, original_id);
        uint32_t decoded_id = GetMsgID(buf);
        EXPECT_EQ(decoded_id, original_id) << "Failed for MsgID: 0x" << std::hex << original_id;
    }
}

// Test PutMsgID big-endian encoding
TEST_F(ProtocolTest, PutMsgIDBigEndian) {
    uint8_t buf[3];
    uint32_t msg_id = 0x123456;

    PutMsgID(buf, msg_id);

    EXPECT_EQ(buf[0], 0x12);
    EXPECT_EQ(buf[1], 0x34);
    EXPECT_EQ(buf[2], 0x56);
}

// Test GetMsgID big-endian decoding
TEST_F(ProtocolTest, GetMsgIDBigEndian) {
    uint8_t buf[3] = {0xAB, 0xCD, 0xEF};

    uint32_t msg_id = GetMsgID(buf);

    EXPECT_EQ(msg_id, 0xABCDEF);
}

// Test NewMessage with empty body
TEST_F(ProtocolTest, NewMessageEmptyBody) {
    std::vector<uint8_t> empty_body;
    auto message = NewMessage(MSG_INVOKE_REQUEST, 12345, empty_body);

    EXPECT_EQ(message.size(), HEADER_SIZE);
    EXPECT_EQ(message[0], VERSION_1);

    uint32_t msg_id = GetMsgID(&message[1]);
    EXPECT_EQ(msg_id, MSG_INVOKE_REQUEST);

    uint32_t req_id = (static_cast<uint32_t>(message[4]) << 24) |
                      (static_cast<uint32_t>(message[5]) << 16) |
                      (static_cast<uint32_t>(message[6]) << 8) |
                      static_cast<uint32_t>(message[7]);
    EXPECT_EQ(req_id, 12345);
}

// Test NewMessage with body
TEST_F(ProtocolTest, NewMessageWithBody) {
    std::vector<uint8_t> body = {0x01, 0x02, 0x03, 0x04, 0x05};
    auto message = NewMessage(MSG_HEARTBEAT_REQUEST, 67890, body);

    EXPECT_EQ(message.size(), HEADER_SIZE + body.size());
    EXPECT_EQ(message[0], VERSION_1);

    // Check body is copied correctly
    bool body_matches = true;
    for (size_t i = 0; i < body.size(); ++i) {
        if (message[HEADER_SIZE + i] != body[i]) {
            body_matches = false;
            break;
        }
    }
    EXPECT_TRUE(body_matches);
}

// Test ParseMessage with valid data
TEST_F(ProtocolTest, ParseMessageValid) {
    std::vector<uint8_t> body = {0xAA, 0xBB, 0xCC};
    auto message = NewMessage(MSG_INVOKE_REQUEST, 11111, body);

    ParsedMessage parsed = ParseMessage(message);

    EXPECT_EQ(parsed.version, VERSION_1);
    EXPECT_EQ(parsed.msg_id, MSG_INVOKE_REQUEST);
    EXPECT_EQ(parsed.req_id, 11111);
    EXPECT_EQ(parsed.body.size(), body.size());
    EXPECT_EQ(parsed.body[0], 0xAA);
    EXPECT_EQ(parsed.body[1], 0xBB);
    EXPECT_EQ(parsed.body[2], 0xCC);
}

// Test ParseMessage with short data
TEST_F(ProtocolTest, ParseMessageTooShort) {
    std::vector<uint8_t> short_data = {0x01, 0x02};  // Less than HEADER_SIZE

    EXPECT_THROW(ParseMessage(short_data), std::runtime_error);
}

// Test ParseMessage with exact header size
TEST_F(ProtocolTest, ParseMessageHeaderOnly) {
    std::vector<uint8_t> header_only_data(HEADER_SIZE);
    header_only_data[0] = VERSION_1;
    PutMsgID(&header_only_data[1], MSG_INVOKE_REQUEST);
    header_only_data[4] = 0x12;  // Request ID big-endian
    header_only_data[5] = 0x34;
    header_only_data[6] = 0x56;
    header_only_data[7] = 0x78;

    ParsedMessage parsed = ParseMessage(header_only_data);

    EXPECT_EQ(parsed.version, VERSION_1);
    EXPECT_EQ(parsed.msg_id, MSG_INVOKE_REQUEST);
    EXPECT_EQ(parsed.req_id, 0x12345678);
    EXPECT_TRUE(parsed.body.empty());
}

// Test IsRequest
TEST_F(ProtocolTest, IsRequest) {
    EXPECT_TRUE(IsRequest(MSG_INVOKE_REQUEST));
    EXPECT_TRUE(IsRequest(MSG_HEARTBEAT_REQUEST));
    EXPECT_TRUE(IsRequest(MSG_REGISTER_REQUEST));
    EXPECT_TRUE(IsRequest(MSG_START_TASK_REQUEST));
    EXPECT_TRUE(IsRequest(MSG_CANCEL_TASK_REQUEST));

    EXPECT_FALSE(IsRequest(MSG_INVOKE_RESPONSE));
    EXPECT_FALSE(IsRequest(MSG_HEARTBEAT_RESPONSE));
}

// Test IsResponse
TEST_F(ProtocolTest, IsResponse) {
    EXPECT_TRUE(IsResponse(MSG_INVOKE_RESPONSE));
    EXPECT_TRUE(IsResponse(MSG_HEARTBEAT_RESPONSE));
    EXPECT_TRUE(IsResponse(MSG_REGISTER_RESPONSE));

    EXPECT_FALSE(IsResponse(MSG_INVOKE_REQUEST));
    EXPECT_FALSE(IsResponse(MSG_HEARTBEAT_REQUEST));
}

// Test IsRequest with special event messages
TEST_F(ProtocolTest, IsRequestEventMessages) {
    // MSG_TASK_EVENT and MSG_METRIC_EVENT are even but not responses
    EXPECT_FALSE(IsRequest(MSG_TASK_EVENT));
    EXPECT_FALSE(IsRequest(MSG_METRIC_EVENT));
}

// Test IsResponse with special event messages
TEST_F(ProtocolTest, IsResponseEventMessages) {
    // MSG_TASK_EVENT and MSG_METRIC_EVENT are even but not responses
    EXPECT_FALSE(IsResponse(MSG_TASK_EVENT));
    EXPECT_FALSE(IsResponse(MSG_METRIC_EVENT));
}

// Test GetResponseMsgID
TEST_F(ProtocolTest, GetResponseMsgID) {
    EXPECT_EQ(GetResponseMsgID(MSG_INVOKE_REQUEST), MSG_INVOKE_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_HEARTBEAT_REQUEST), MSG_HEARTBEAT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_REGISTER_REQUEST), MSG_REGISTER_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_START_TASK_REQUEST), MSG_START_TASK_RESPONSE);
}

// Test MsgIDString for known messages
TEST_F(ProtocolTest, MsgIDStringKnown) {
    EXPECT_EQ(MsgIDString(MSG_INVOKE_REQUEST), "InvokeRequest");
    EXPECT_EQ(MsgIDString(MSG_INVOKE_RESPONSE), "InvokeResponse");
    EXPECT_EQ(MsgIDString(MSG_HEARTBEAT_REQUEST), "HeartbeatRequest");
    EXPECT_EQ(MsgIDString(MSG_HEARTBEAT_RESPONSE), "HeartbeatResponse");
    EXPECT_EQ(MsgIDString(MSG_REGISTER_REQUEST), "RegisterRequest");
    EXPECT_EQ(MsgIDString(MSG_REGISTER_RESPONSE), "RegisterResponse");
    EXPECT_EQ(MsgIDString(MSG_START_TASK_REQUEST), "StartTaskRequest");
    EXPECT_EQ(MsgIDString(MSG_TASK_EVENT), "TaskEvent");
    EXPECT_EQ(MsgIDString(MSG_PROVIDER_CONNECT_REQUEST), "ProviderConnectRequest");
}

// Test MsgIDString for unknown messages
TEST_F(ProtocolTest, MsgIDStringUnknown) {
    std::string unknown = MsgIDString(0xDEADBEEF);
    EXPECT_NE(unknown.find("Unknown"), std::string::npos);
    EXPECT_NE(unknown.find("0xDEADBEEF"), std::string::npos);
}

// Test request ID big-endian encoding in NewMessage
TEST_F(ProtocolTest, NewMessageRequestIDBigEndian) {
    auto message = NewMessage(MSG_INVOKE_REQUEST, 0x12345678, {});

    EXPECT_EQ(message[4], 0x12);
    EXPECT_EQ(message[5], 0x34);
    EXPECT_EQ(message[6], 0x56);
    EXPECT_EQ(message[7], 0x78);
}

// Test request ID big-endian decoding in ParseMessage
TEST_F(ProtocolTest, ParseMessageRequestIDBigEndian) {
    std::vector<uint8_t> data(HEADER_SIZE);
    data[0] = VERSION_1;
    PutMsgID(&data[1], MSG_INVOKE_REQUEST);
    data[4] = 0xAB;
    data[5] = 0xCD;
    data[6] = 0xEF;
    data[7] = 0x01;

    ParsedMessage parsed = ParseMessage(data);

    EXPECT_EQ(parsed.req_id, 0xABCDEF01);
}

// Test message round-trip
TEST_F(ProtocolTest, MessageRoundTrip) {
    std::vector<uint8_t> original_body = {0xDE, 0xAD, 0xBE, 0xEF};
    uint32_t original_msg_id = MSG_INVOKE_REQUEST;
    uint32_t original_req_id = 0x12345678;

    auto message = NewMessage(original_msg_id, original_req_id, original_body);
    ParsedMessage parsed = ParseMessage(message);

    EXPECT_EQ(parsed.msg_id, original_msg_id);
    EXPECT_EQ(parsed.req_id, original_req_id);
    EXPECT_EQ(parsed.body, original_body);
}

// Test MSG_TASK_EVENT and MSG_METRIC_EVENT special handling
TEST_F(ProtocolTest, EventMessagesSpecialCase) {
    // These are event messages, not requests or responses
    EXPECT_FALSE(IsRequest(MSG_TASK_EVENT));
    EXPECT_FALSE(IsResponse(MSG_TASK_EVENT));

    EXPECT_FALSE(IsRequest(MSG_METRIC_EVENT));
    EXPECT_FALSE(IsResponse(MSG_METRIC_EVENT));
}

// Test ControlService message IDs
TEST_F(ProtocolTest, ControlServiceMessageIDs) {
    EXPECT_EQ(GetResponseMsgID(MSG_REGISTER_REQUEST), MSG_REGISTER_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_HEARTBEAT_REQUEST), MSG_HEARTBEAT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_REGISTER_CAPABILITIES_REQ), MSG_REGISTER_CAPABILITIES_RESP);
}

// Test ClientService message IDs
TEST_F(ProtocolTest, ClientServiceMessageIDs) {
    EXPECT_EQ(GetResponseMsgID(MSG_REGISTER_CLIENT_REQUEST), MSG_REGISTER_CLIENT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_CLIENT_HEARTBEAT_REQUEST), MSG_CLIENT_HEARTBEAT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_LIST_CLIENTS_REQUEST), MSG_LIST_CLIENTS_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_GET_TASK_RESULT_REQUEST), MSG_GET_TASK_RESULT_RESPONSE);
}

// Test InvokerService message IDs
TEST_F(ProtocolTest, InvokerServiceMessageIDs) {
    EXPECT_EQ(GetResponseMsgID(MSG_INVOKE_REQUEST), MSG_INVOKE_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_START_TASK_REQUEST), MSG_START_TASK_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_CANCEL_TASK_REQUEST), MSG_CANCEL_TASK_RESPONSE);
}

// Test OpsService message IDs
TEST_F(ProtocolTest, OpsServiceMessageIDs) {
    EXPECT_EQ(GetResponseMsgID(MSG_GET_SYSTEM_INFO_REQUEST), MSG_GET_SYSTEM_INFO_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_LIST_PROCESSES_REQUEST), MSG_LIST_PROCESSES_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_REPORT_METRICS_REQUEST), MSG_REPORT_METRICS_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_RESTART_PROCESS_REQUEST), MSG_RESTART_PROCESS_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_STOP_PROCESS_REQUEST), MSG_STOP_PROCESS_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_START_PROCESS_REQUEST), MSG_START_PROCESS_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_EXECUTE_COMMAND_REQUEST), MSG_EXECUTE_COMMAND_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_LIST_SERVICES_REQUEST), MSG_LIST_SERVICES_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_GET_SERVICE_STATUS_REQUEST), MSG_GET_SERVICE_STATUS_RESPONSE);
}

// Test ProviderSession message IDs
TEST_F(ProtocolTest, ProviderSessionMessageIDs) {
    EXPECT_EQ(GetResponseMsgID(MSG_PROVIDER_CONNECT_REQUEST), MSG_PROVIDER_CONNECT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_PROVIDER_HEARTBEAT_REQUEST), MSG_PROVIDER_HEARTBEAT_RESPONSE);
    EXPECT_EQ(GetResponseMsgID(MSG_PROVIDER_DRAIN_REQUEST), MSG_PROVIDER_DRAIN_RESPONSE);
}

// Test HeaderSize constant
TEST_F(ProtocolTest, HeaderSizeConstant) {
    // Version(1) + MsgID(3) + RequestID(4) = 8
    EXPECT_EQ(HEADER_SIZE, 8);
}

// Test VERSION_1 constant
TEST_F(ProtocolTest, VersionConstant) {
    EXPECT_EQ(VERSION_1, 0x01);
}

// Test NewMessage with zero request ID
TEST_F(ProtocolTest, NewMessageZeroRequestID) {
    auto message = NewMessage(MSG_INVOKE_REQUEST, 0, {});

    uint32_t req_id = (static_cast<uint32_t>(message[4]) << 24) |
                      (static_cast<uint32_t>(message[5]) << 16) |
                      (static_cast<uint32_t>(message[6]) << 8) |
                      static_cast<uint32_t>(message[7]);
    EXPECT_EQ(req_id, 0);
}

// Test NewMessage with maximum request ID
TEST_F(ProtocolTest, NewMessageMaxRequestID) {
    uint32_t max_req_id = 0xFFFFFFFF;
    auto message = NewMessage(MSG_INVOKE_REQUEST, max_req_id, {});

    uint32_t req_id = (static_cast<uint32_t>(message[4]) << 24) |
                      (static_cast<uint32_t>(message[5]) << 16) |
                      (static_cast<uint32_t>(message[6]) << 8) |
                      static_cast<uint32_t>(message[7]);
    EXPECT_EQ(req_id, max_req_id);
}

// Test ParseMessage with large body
TEST_F(ProtocolTest, ParseMessageLargeBody) {
    std::vector<uint8_t> large_body(10000, 0xAB);
    auto message = NewMessage(MSG_INVOKE_REQUEST, 123, large_body);

    ParsedMessage parsed = ParseMessage(message);

    EXPECT_EQ(parsed.body.size(), 10000);
    EXPECT_TRUE(std::all_of(parsed.body.begin(), parsed.body.end(),
                            [](uint8_t b) { return b == 0xAB; }));
}

// Test MsgIDString for all known messages
TEST_F(ProtocolTest, MsgIDStringAllKnown) {
    // Test that all known messages return non-empty strings
    std::vector<uint32_t> known_msgs = {
        MSG_REGISTER_REQUEST, MSG_REGISTER_RESPONSE,
        MSG_HEARTBEAT_REQUEST, MSG_HEARTBEAT_RESPONSE,
        MSG_REGISTER_CAPABILITIES_REQ, MSG_REGISTER_CAPABILITIES_RESP,
        MSG_REGISTER_CLIENT_REQUEST, MSG_REGISTER_CLIENT_RESPONSE,
        MSG_CLIENT_HEARTBEAT_REQUEST, MSG_CLIENT_HEARTBEAT_RESPONSE,
        MSG_LIST_CLIENTS_REQUEST, MSG_LIST_CLIENTS_RESPONSE,
        MSG_GET_TASK_RESULT_REQUEST, MSG_GET_TASK_RESULT_RESPONSE,
        MSG_INVOKE_REQUEST, MSG_INVOKE_RESPONSE,
        MSG_START_TASK_REQUEST, MSG_START_TASK_RESPONSE,
        MSG_STREAM_TASK_REQUEST, MSG_TASK_EVENT,
        MSG_CANCEL_TASK_REQUEST, MSG_CANCEL_TASK_RESPONSE,
        MSG_GET_SYSTEM_INFO_REQUEST, MSG_GET_SYSTEM_INFO_RESPONSE,
        MSG_LIST_PROCESSES_REQUEST, MSG_LIST_PROCESSES_RESPONSE,
        MSG_REPORT_METRICS_REQUEST, MSG_REPORT_METRICS_RESPONSE,
        MSG_STREAM_METRICS_REQUEST, MSG_METRIC_EVENT,
        MSG_RESTART_PROCESS_REQUEST, MSG_RESTART_PROCESS_RESPONSE,
        MSG_STOP_PROCESS_REQUEST, MSG_STOP_PROCESS_RESPONSE,
        MSG_START_PROCESS_REQUEST, MSG_START_PROCESS_RESPONSE,
        MSG_EXECUTE_COMMAND_REQUEST, MSG_EXECUTE_COMMAND_RESPONSE,
        MSG_LIST_SERVICES_REQUEST, MSG_LIST_SERVICES_RESPONSE,
        MSG_GET_SERVICE_STATUS_REQUEST, MSG_GET_SERVICE_STATUS_RESPONSE,
        MSG_PROVIDER_CONNECT_REQUEST, MSG_PROVIDER_CONNECT_RESPONSE,
        MSG_PROVIDER_HEARTBEAT_REQUEST, MSG_PROVIDER_HEARTBEAT_RESPONSE,
        MSG_PROVIDER_DRAIN_REQUEST, MSG_PROVIDER_DRAIN_RESPONSE
    };

    for (uint32_t msg_id : known_msgs) {
        std::string str = MsgIDString(msg_id);
        EXPECT_FALSE(str.empty()) << "Empty string for MsgID: 0x" << std::hex << msg_id;
        EXPECT_EQ(str.find("Unknown"), std::string::npos) << "Unknown string for MsgID: 0x" << std::hex << msg_id;
    }
}

// Test odd/even MsgID pattern
TEST_F(ProtocolTest, OddEvenMsgIDPattern) {
    // All requests should have odd MsgID
    EXPECT_TRUE(MSG_INVOKE_REQUEST % 2 == 1);
    EXPECT_TRUE(MSG_HEARTBEAT_REQUEST % 2 == 1);
    EXPECT_TRUE(MSG_REGISTER_REQUEST % 2 == 1);

    // All responses should have even MsgID
    EXPECT_TRUE(MSG_INVOKE_RESPONSE % 2 == 0);
    EXPECT_TRUE(MSG_HEARTBEAT_RESPONSE % 2 == 0);
    EXPECT_TRUE(MSG_REGISTER_RESPONSE % 2 == 0);
}

}  // namespace test
}  // namespace protocol
}  // namespace sdk
}  // namespace croupier

File: sdks/python/tests/test_protocol.py
"""Tests for Croupier wire protocol implementation."""

import struct

import pytest

from croupier import protocol


class TestProtocolConstants:
    """Test protocol constants and message type definitions."""

    def test_version_constant(self):
        assert protocol.VERSION_1 == 0x01

    def test_header_size(self):
        assert protocol.HEADER_SIZE == 8

    def test_message_type_constants_are_unique(self):
        """All message type constants should be unique."""
        msg_types = [
            protocol.MSG_REGISTER_REQUEST,
            protocol.MSG_REGISTER_RESPONSE,
            protocol.MSG_HEARTBEAT_REQUEST,
            protocol.MSG_HEARTBEAT_RESPONSE,
            protocol.MSG_INVOKE_REQUEST,
            protocol.MSG_INVOKE_RESPONSE,
            protocol.MSG_START_TASK_REQUEST,
            protocol.MSG_START_TASK_RESPONSE,
            protocol.MSG_TASK_EVENT,
            protocol.MSG_CANCEL_TASK_REQUEST,
            protocol.MSG_CANCEL_TASK_RESPONSE,
            protocol.MSG_PROVIDER_CONNECT_REQUEST,
            protocol.MSG_PROVIDER_CONNECT_RESPONSE,
            protocol.MSG_PROVIDER_HEARTBEAT_REQUEST,
            protocol.MSG_PROVIDER_HEARTBEAT_RESPONSE,
        ]
        assert len(msg_types) == len(set(msg_types))

    def test_request_ids_are_odd(self):
        """Request message IDs should be odd."""
        request_ids = [
            protocol.MSG_REGISTER_REQUEST,
            protocol.MSG_HEARTBEAT_REQUEST,
            protocol.MSG_INVOKE_REQUEST,
            protocol.MSG_START_TASK_REQUEST,
            protocol.MSG_CANCEL_TASK_REQUEST,
            protocol.MSG_PROVIDER_CONNECT_REQUEST,
            protocol.MSG_PROVIDER_HEARTBEAT_REQUEST,
        ]
        for msg_id in request_ids:
            assert msg_id % 2 == 1, f"Request ID 0x{msg_id:06X} should be odd"

    def test_response_ids_are_even(self):
        """Response message IDs should be even."""
        response_ids = [
            protocol.MSG_REGISTER_RESPONSE,
            protocol.MSG_HEARTBEAT_RESPONSE,
            protocol.MSG_INVOKE_RESPONSE,
            protocol.MSG_START_TASK_RESPONSE,
            protocol.MSG_CANCEL_TASK_RESPONSE,
            protocol.MSG_PROVIDER_CONNECT_RESPONSE,
            protocol.MSG_PROVIDER_HEARTBEAT_RESPONSE,
        ]
        for msg_id in response_ids:
            assert msg_id % 2 == 0, f"Response ID 0x{msg_id:06X} should be even"


class TestPutMsgId:
    """Test put_msg_id function."""

    def test_basic_encoding(self):
        msg_id = 0x030101
        result = protocol.put_msg_id(msg_id)
        assert len(result) == 3
        assert result == bytes([0x03, 0x01, 0x01])

    def test_max_value(self):
        msg_id = 0xFFFFFF
        result = protocol.put_msg_id(msg_id)
        assert result == bytes([0xFF, 0xFF, 0xFF])

    def test_zero(self):
        result = protocol.put_msg_id(0)
        assert result == bytes([0x00, 0x00, 0x00])

    def test_single_byte_values(self):
        assert protocol.put_msg_id(0x01) == bytes([0x00, 0x00, 0x01])
        assert protocol.put_msg_id(0x100) == bytes([0x00, 0x01, 0x00])
        assert protocol.put_msg_id(0x10000) == bytes([0x01, 0x00, 0x00])


class TestGetMsgId:
    """Test get_msg_id function."""

    def test_basic_decoding(self):
        data = bytes([0x03, 0x01, 0x01])
        assert protocol.get_msg_id(data) == 0x030101

    def test_max_value(self):
        data = bytes([0xFF, 0xFF, 0xFF])
        assert protocol.get_msg_id(data) == 0xFFFFFF

    def test_zero(self):
        data = bytes([0x00, 0x00, 0x00])
        assert protocol.get_msg_id(data) == 0

    def test_roundtrip(self):
        """Encoding then decoding should return original value."""
        for msg_id in [0x010101, 0x030101, 0xFFFFFF, 0x000001]:
            assert protocol.get_msg_id(protocol.put_msg_id(msg_id)) == msg_id


class TestNewMessage:
    """Test new_message function."""

    def test_basic_message(self):
        msg_id = protocol.MSG_INVOKE_REQUEST
        req_id = 12345
        body = b"test payload"

        message = protocol.new_message(msg_id, req_id, body)

        assert len(message) == protocol.HEADER_SIZE + len(body)
        assert message[0] == protocol.VERSION_1

    def test_empty_body(self):
        msg_id = protocol.MSG_HEARTBEAT_REQUEST
        req_id = 999

        message = protocol.new_message(msg_id, req_id, b"")

        assert len(message) == protocol.HEADER_SIZE

    def test_none_body(self):
        """None body should be treated as empty bytes."""
        msg_id = protocol.MSG_HEARTBEAT_REQUEST
        req_id = 999

        message = protocol.new_message(msg_id, req_id, b"")

        assert len(message) == protocol.HEADER_SIZE

    def test_large_body(self):
        msg_id = protocol.MSG_INVOKE_REQUEST
        req_id = 1
        body = b"x" * 1024 * 1024  # 1MB

        message = protocol.new_message(msg_id, req_id, body)

        assert len(message) == protocol.HEADER_SIZE + len(body)

    def test_message_structure(self):
        """Verify the exact structure of a message."""
        msg_id = 0x030101
        req_id = 0x00000001
        body = b"hello"

        message = protocol.new_message(msg_id, req_id, body)

        # Version byte
        assert message[0] == 0x01
        # MsgID (3 bytes, big-endian)
        assert message[1:4] == bytes([0x03, 0x01, 0x01])
        # RequestID (4 bytes, big-endian)
        assert message[4:8] == struct.pack(">I", 1)
        # Body
        assert message[8:] == b"hello"


class TestParseMessage:
    """Test parse_message function."""

    def test_basic_parse(self):
        msg_id = protocol.MSG_INVOKE_REQUEST
        req_id = 999
        body = b"hello world"

        message = protocol.new_message(msg_id, req_id, body)
        version, parsed_msg_id, parsed_req_id, parsed_body = protocol.parse_message(message)

        assert version == protocol.VERSION_1
        assert parsed_msg_id == msg_id
        assert parsed_req_id == req_id
        assert parsed_body == body

    def test_empty_body(self):
        msg_id = protocol.MSG_HEARTBEAT_REQUEST
        req_id = 12345

        message = protocol.new_message(msg_id, req_id, b"")
        version, parsed_msg_id, parsed_req_id, parsed_body = protocol.parse_message(message)

        assert version == protocol.VERSION_1
        assert parsed_msg_id == msg_id
        assert parsed_req_id == req_id
        assert parsed_body == b""

    def test_too_short_raises_error(self):
        short_message = bytes([0x01, 0x02, 0x03])
        with pytest.raises(ValueError, match="Message too short"):
            protocol.parse_message(short_message)

    def test_exactly_header_size(self):
        """Parsing a message with exactly header size should return empty body."""
        msg_id = protocol.MSG_HEARTBEAT_REQUEST
        req_id = 1

        message = protocol.new_message(msg_id, req_id, b"")
        assert len(message) == protocol.HEADER_SIZE

        version, parsed_msg_id, parsed_req_id, parsed_body = protocol.parse_message(message)
        assert parsed_body == b""

    def test_roundtrip(self):
        """Creating and parsing should return original values."""
        msg_id = protocol.MSG_REGISTER_REQUEST
        req_id = 42
        body = b"test data"

        message = protocol.new_message(msg_id, req_id, body)
        version, parsed_msg_id, parsed_req_id, parsed_body = protocol.parse_message(message)

        assert parsed_msg_id == msg_id
        assert parsed_req_id == req_id
        assert parsed_body == body


class TestIsRequest:
    """Test is_request function."""

    def test_request_messages(self):
        assert protocol.is_request(protocol.MSG_REGISTER_REQUEST) is True
        assert protocol.is_request(protocol.MSG_HEARTBEAT_REQUEST) is True
        assert protocol.is_request(protocol.MSG_INVOKE_REQUEST) is True
        assert protocol.is_request(protocol.MSG_START_TASK_REQUEST) is True
        assert protocol.is_request(protocol.MSG_CANCEL_TASK_REQUEST) is True
        assert protocol.is_request(protocol.MSG_PROVIDER_CONNECT_REQUEST) is True
        assert protocol.is_request(protocol.MSG_PROVIDER_HEARTBEAT_REQUEST) is True

    def test_response_messages(self):
        assert protocol.is_request(protocol.MSG_REGISTER_RESPONSE) is False
        assert protocol.is_request(protocol.MSG_HEARTBEAT_RESPONSE) is False
        assert protocol.is_request(protocol.MSG_INVOKE_RESPONSE) is False

    def test_event_messages(self):
        """TaskEvent and MetricEvent are neither requests nor responses."""
        assert protocol.is_request(protocol.MSG_TASK_EVENT) is False
        assert protocol.is_request(protocol.MSG_METRIC_EVENT) is False


class TestIsResponse:
    """Test is_response function."""

    def test_response_messages(self):
        assert protocol.is_response(protocol.MSG_REGISTER_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_HEARTBEAT_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_INVOKE_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_START_TASK_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_CANCEL_TASK_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_PROVIDER_CONNECT_RESPONSE) is True
        assert protocol.is_response(protocol.MSG_PROVIDER_HEARTBEAT_RESPONSE) is True

    def test_request_messages(self):
        assert protocol.is_response(protocol.MSG_REGISTER_REQUEST) is False
        assert protocol.is_response(protocol.MSG_HEARTBEAT_REQUEST) is False
        assert protocol.is_response(protocol.MSG_INVOKE_REQUEST) is False

    def test_event_messages(self):
        """TaskEvent and MetricEvent are neither requests nor responses."""
        assert protocol.is_response(protocol.MSG_TASK_EVENT) is False
        assert protocol.is_response(protocol.MSG_METRIC_EVENT) is False


class TestGetResponseMsgId:
    """Test get_response_msg_id function."""

    def test_basic_pairs(self):
        assert protocol.get_response_msg_id(protocol.MSG_REGISTER_REQUEST) == protocol.MSG_REGISTER_RESPONSE
        assert protocol.get_response_msg_id(protocol.MSG_HEARTBEAT_REQUEST) == protocol.MSG_HEARTBEAT_RESPONSE
        assert protocol.get_response_msg_id(protocol.MSG_INVOKE_REQUEST) == protocol.MSG_INVOKE_RESPONSE
        assert protocol.get_response_msg_id(protocol.MSG_START_TASK_REQUEST) == protocol.MSG_START_TASK_RESPONSE
        assert protocol.get_response_msg_id(protocol.MSG_CANCEL_TASK_REQUEST) == protocol.MSG_CANCEL_TASK_RESPONSE
        assert protocol.get_response_msg_id(protocol.MSG_PROVIDER_CONNECT_REQUEST) == protocol.MSG_PROVIDER_CONNECT_RESPONSE

    def test_response_is_request_plus_one(self):
        """Response ID should be request ID + 1."""
        request_ids = [
            protocol.MSG_REGISTER_REQUEST,
            protocol.MSG_HEARTBEAT_REQUEST,
            protocol.MSG_INVOKE_REQUEST,
            protocol.MSG_START_TASK_REQUEST,
            protocol.MSG_CANCEL_TASK_REQUEST,
            protocol.MSG_PROVIDER_CONNECT_REQUEST,
            protocol.MSG_PROVIDER_HEARTBEAT_REQUEST,
        ]
        for req_id in request_ids:
            assert protocol.get_response_msg_id(req_id) == req_id + 1


class TestMsgIdString:
    """Test msg_id_string function."""

    def test_known_message_types(self):
        assert protocol.msg_id_string(protocol.MSG_REGISTER_REQUEST) == "RegisterRequest"
        assert protocol.msg_id_string(protocol.MSG_REGISTER_RESPONSE) == "RegisterResponse"
        assert protocol.msg_id_string(protocol.MSG_HEARTBEAT_REQUEST) == "HeartbeatRequest"
        assert protocol.msg_id_string(protocol.MSG_HEARTBEAT_RESPONSE) == "HeartbeatResponse"
        assert protocol.msg_id_string(protocol.MSG_INVOKE_REQUEST) == "InvokeRequest"
        assert protocol.msg_id_string(protocol.MSG_INVOKE_RESPONSE) == "InvokeResponse"
        assert protocol.msg_id_string(protocol.MSG_START_TASK_REQUEST) == "StartTaskRequest"
        assert protocol.msg_id_string(protocol.MSG_TASK_EVENT) == "TaskEvent"
        assert protocol.msg_id_string(protocol.MSG_PROVIDER_CONNECT_REQUEST) == "ProviderConnectRequest"

    def test_unknown_message_type(self):
        result = protocol.msg_id_string(0xFFFFFF)
        assert result.startswith("Unknown")
        assert "0xFFFFFF" in result

    def test_zero(self):
        result = protocol.msg_id_string(0)
        assert result.startswith("Unknown")


class TestRequestResponsePairs:
    """Test that request/response pairs are correctly defined."""

    @pytest.mark.parametrize(
        "request_id,response_id",
        [
            (protocol.MSG_REGISTER_REQUEST, protocol.MSG_REGISTER_RESPONSE),
            (protocol.MSG_HEARTBEAT_REQUEST, protocol.MSG_HEARTBEAT_RESPONSE),
            (protocol.MSG_INVOKE_REQUEST, protocol.MSG_INVOKE_RESPONSE),
            (protocol.MSG_START_TASK_REQUEST, protocol.MSG_START_TASK_RESPONSE),
            (protocol.MSG_CANCEL_TASK_REQUEST, protocol.MSG_CANCEL_TASK_RESPONSE),
            (protocol.MSG_PROVIDER_CONNECT_REQUEST, protocol.MSG_PROVIDER_CONNECT_RESPONSE),
            (protocol.MSG_PROVIDER_HEARTBEAT_REQUEST, protocol.MSG_PROVIDER_HEARTBEAT_RESPONSE),
        ],
    )
    def test_pair_consistency(self, request_id, response_id):
        """Request and response IDs should be consistent."""
        assert response_id == protocol.get_response_msg_id(request_id)
        assert protocol.is_request(request_id) is True
        assert protocol.is_response(response_id) is True
        assert protocol.is_request(response_id) is False
        assert protocol.is_response(request_id) is False
